courses

Tu-S opened the pull request:
Знаков поворота нет, так что из любого IN можно доехать в любой OUT. Пересечения считаются, когда потоки машин из разных IN заезжают в одну и ту же полосу OUT.  Внутри перекрестка пересечения не считаются.

Tu-S committed:
courses

Знаков поворота нет, так что из любого IN можно доехать в любой OUT. Пересечения считаются, когда потоки машин из разных IN заезжают в одну и ту же полосу OUT.  Внутри перекрестка пересечения не считаются.

Tu-S committed:
fixed mistake

SpykeX3 commented on Engine/src/main/java/ru/nsu/team/Main.java:
```diff
@@ -9,15 +10,15 @@ public static void main(String[] args) {
-        Master master = new Master();
-        master.execute(args);
+        SimulationController simulationController = new SimulationController();
+        simulationController.testRun("config/test.tsp");
```
Давай не ломать main. Если нужно написать тесты, то для этого можно использовать тесты)

SpykeX3 commented on Engine/src/main/java/ru/nsu/team/roadmodelcreator/RoadModelCreator.java:
```diff
@@ -115,22 +114,69 @@ private void createCourses(List<RoadConfiguration> roadsConfig, List<NodeConfigu
+                                    var intersection = new Intersection(5);
+                                    crFromLane.addIntersection(intersection);
+                                    cr.addIntersection(intersection);
+                                }
```
В эту часть пока не вникал, но давай сделаем, чтобы на перекрестке без светофора машины не могли сквозь перпендикулярный поток проезжать

Tu-S replied:
![image](https://user-images.githubusercontent.com/42936486/115980660-a07c5380-a5b8-11eb-9827-3f2a28775645.png)
Если я тебя правильно понял, то для синих зеленый и оранжевый потоки перпендикулярны, а куда им тогда ехать? В таких случаях обязательно должен быть светофор, как мне кажется. Это с точки зрения дороги какой то косяк уже.

SpykeX3 replied:
![intersection](https://user-images.githubusercontent.com/16012150/116290859-21f10300-a7be-11eb-91d7-a03dcd416191.png)
Я про то, что если едут оранжевые и зеленые, то фиолетовые и синие не могут проехать. То есть давай сделаем intersection общий для курса (зеленых и синих), (оранжевых и синих), (фиолетовых и зеленых), (фиолетовых и оранжевых)

Tu-S replied:
Я понял, что ты хочешь intersectaion для синих и зеленых, чтобы когда зеленые едут из 8 в 6, синие из 5 не могли проехать в 9 (в 6 они не проедут, потому что на этот случай создается intersaction). Вот что делать с 9 в этом случае, я не придумал, но если мы запускаем зеленых и оранжевых, то в 9 возникает конфликт синих и оранжевых, так что синие ехать не смогут. Intersaction создается если два курса ведут в одну дорогу, более того в одну полосу. Как определять их пересечения внутри самой node, повторю, я не придумал. Мне кажется это не критично. В любом случае мы добиваемся того, что два потока не могу одновременно ломиться в один выход из node.   
На картинке меньше дорог, но сути не меняет, просто в конфигах нет случая для 4 дорог, но логика будет та же  
![image](https://user-images.githubusercontent.com/42936486/116347110-c5213700-a815-11eb-8aec-58d9ba27b456.png)

SpykeX3 replied:
Пересечение курсов (8,6) и (5,9) важное, без таких пересечений у нас не только с 3 дорогами будет кривовато работать симуляция, но и вообще неправильно себя вести перекресток 4 дорог. Если совсем не можешь так сделать, то можем смержить как есть и я этим займусь

SpykeX3 commented on Engine/src/main/java/ru/nsu/team/roadmodelcreator/RoadModelCreator.java:
```diff
@@ -115,22 +114,69 @@ private void createCourses(List<RoadConfiguration> roadsConfig, List<NodeConfigu
+        for (var fI : f.getIntersections()) {
+            for (var sI : s.getIntersections()) {
+                if (fI.equals(sI)) {
+                    return true;
```
Тут получается цикл вложенности 3, мб можно сеты и пересечения можно заюзать?
(Не критично)

SpykeX3 commented on Engine/src/main/java/ru/nsu/team/roadmodelcreator/RoadModelCreator.java:
```diff
@@ -283,5 +329,15 @@ private void analyzeSign(SignConfiguration sign, Lane lane) {
+        List<Course> courses;
+
+        public KeyValue(Lane key, List<Course> courses) {
+            this.key = key;
```
Давай здесь использовать обычную javax.что-то.Pair

Tu-S replied:
javafx.util.Pair<K,V>, сейчас не входил в основной jdk, не думаю, что стоит тащить весь fx ради 1 класса. Хорошая новость,  я увидел класс, который уже сделал Вова.

SpykeX3 replied:
Сорян, забыл что это javafx, думал, что из javax. Сейчас выглядит норм.

Tu-S committed:
fixes

Tu-S committed:
upd

Сейчас пересечения создаются со всеми курсами, а не только с теми, которые ведут в одну дорогу. Таким образом, мы получаем нужные пересечения синих и зеленых, синих и оранжевых.

Tu-S committed:
removed unnecessary "if" conditions

Tu-S committed:
добавил курс  любому направлению

Tu-S committed:
intersaction для любого курса
